VPLAY-9274:[DASH] optimization: independent track downloads

nu641001 opened the pull request:
Reason for change: Implementing independent track downloads in dash downloads
Risks: Low
Test Procedure: Test with all sort of DASH contents
Priority: P1

nu641001 committed:
VPLAY-9274:[DASH] optimization: independent track downloads

Reason for change: Implementing independent track downloads in dash downloads
Risks: Low
Test Procedure: Test with all sort of DASH contents
Priority: P1
Signed-off-by: Nandakishor Udiyannur Mana <nu641001@gmail.com>

rdkcmf-jenkins commented on dash/mpd/MPDModel.cpp:
```diff
@@ -179,6 +179,97 @@ string findBaseUrl(DomElement &element, const string &current, bool isFile) {
+        string text = eUrl.text();
+        if (!text.empty())
+        {
+            baseUrlTexts.push_back(text);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
4320033efac6c09eb3e635072e8c6969
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"text" is copied and then passed-by-reference as parameter to STL insertion function "std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::push_back(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::value_type const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""text"")" instead of "text".

</details>

nu641001 replied:
Resolved

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -14296,11 +13814,270 @@ void StreamAbstractionAAMP_MPD::SetSubtitleTrackOffset()
+				// For non-fog TSB, we can use the representation's bandwidth
+				// to construct the URL for init fragments
+				ConstructFragmentURL(uriInfo.url, fragmentDescriptor.get(), std::move(urlTemplate), aamp->mConfig);
+				uriList[fragmentDescriptor->Bandwidth] = uriInfo;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
ffdfc7f0f8e995f31377f1924a8522eb
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"uriInfo" is copied in call to copy assignment for class "URIInfo", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""uriInfo"")" instead of "uriInfo".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -14296,11 +13814,270 @@ void StreamAbstractionAAMP_MPD::SetSubtitleTrackOffset()
+			// For segment base, we need to use the index range for the URL
+			// The range is identified by parsing sidx box and it is done at downloader level
+			ConstructFragmentURL(uriInfo.url, fragmentDescriptor.get(), "", aamp->mConfig);
+			uriList[fragmentDescriptor->Bandwidth] = uriInfo;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
ffdfc7f0f8e995f31377f1924a8522eb
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"uriInfo" is copied in call to copy assignment for class "URIInfo", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""uriInfo"")" instead of "uriInfo".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -14296,11 +13814,270 @@ void StreamAbstractionAAMP_MPD::SetSubtitleTrackOffset()
+				if (!initUrl.empty())
+				{
+					ConstructFragmentURL(uriInfo.url, fragmentDescriptor.get(), std::move(initUrl), aamp->mConfig);
+					uriList[fragmentDescriptor->Bandwidth] = uriInfo;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
ffdfc7f0f8e995f31377f1924a8522eb
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"uriInfo" is copied in call to copy assignment for class "URIInfo", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""uriInfo"")" instead of "uriInfo".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -14296,11 +13814,270 @@ void StreamAbstractionAAMP_MPD::SetSubtitleTrackOffset()
+							snprintf(temp, sizeof(temp), "0-%d", start - 1);
+							uriInfo.range = temp;
+							ConstructFragmentURL(uriInfo.url, fragmentDescriptor.get(), "", aamp->mConfig);
+							uriList[fragmentDescriptor->Bandwidth] = uriInfo;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
ffdfc7f0f8e995f31377f1924a8522eb
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"uriInfo" is copied in call to copy assignment for class "URIInfo", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""uriInfo"")" instead of "uriInfo".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -14296,11 +13814,270 @@ void StreamAbstractionAAMP_MPD::SetSubtitleTrackOffset()
+					{
+						ConstructFragmentURL(uriInfo.url, fragmentDescriptor.get(), segmentURL->GetMediaURI(), aamp->mConfig);
+						uriInfo.range = segmentURL->GetMediaRange();
+						uriList[fragmentDescriptor->Bandwidth] = uriInfo;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
ffdfc7f0f8e995f31377f1924a8522eb
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"uriInfo" is copied in call to copy assignment for class "URIInfo", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""uriInfo"")" instead of "uriInfo".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -14296,11 +13814,270 @@ void StreamAbstractionAAMP_MPD::SetSubtitleTrackOffset()
+					URIInfo fogUriInfo;
+					fragmentDescriptor->Bandwidth = rep->GetBandwidth();
+					ConstructFragmentURL(fogUriInfo.url, fragmentDescriptor.get(), std::move(urlTemplate), aamp->mConfig);
+					uriList[fragmentDescriptor->Bandwidth] = fogUriInfo;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
5520b3522ad35fad07fbcca1fc9478b4
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"fogUriInfo" is copied in call to copy assignment for class "URIInfo", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""fogUriInfo"")" instead of "fogUriInfo".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -14296,11 +13814,270 @@ void StreamAbstractionAAMP_MPD::SetSubtitleTrackOffset()
+				{
+					URIInfo fogUriInfo;
+					fragmentDescriptor->Bandwidth = rep->GetBandwidth();
+					ConstructFragmentURL(fogUriInfo.url, fragmentDescriptor.get(), std::move(urlTemplate), aamp->mConfig);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
98b1e938625ea08cbb80b1c74f0ec0b3
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Using a moved object
"urlTemplate" is used after it has been already moved.

_High Impact, CWE-457_
_USE_AFTER_MOVE_



</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8293,9 +7913,18 @@ double StreamAbstractionAAMP_MPD::GetCulledSeconds(std::vector<PeriodInfo> &curr
+						}
 						if (mMPDParseHelper->GetLiveTimeFragmentSync())
 						{
 							 newStartSegment += (long)((mMPDParseHelper->GetPeriodStartTime(0,mLastPlaylistDownloadTimeMs) - mAvailabilityStartTime) / fragmentDuration);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
c850df35a4cb7dfe22e050b459c251b1
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Division or modulo by float zero
In expression "(this->mMPDParseHelper->GetPeriodStartTime(0, this->mLastPlaylistDownloadTimeMs) - this->mAvailabilityStartTime) / fragmentDuration", division by expression "fragmentDuration" which may be zero has undefined behavior.

_Medium Impact, CWE-369_
_DIVIDE_BY_ZERO_



</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8659,23 +8272,19 @@ void StreamAbstractionAAMP_MPD::FetchAndInjectInitialization(int trackIdx, bool
-								pMediaStreamContext->fragmentTime,
-								0, // duration - zero for init fragment
-								range.c_str(), true ))
+							if (!FetchFragment(pMediaStreamContext, fragmentUrl, 0.0, true, getCurlInstanceByMediaType(pMediaStreamContext->mediaType), false, pMediaStreamContext->discontinuity, 0, pMediaStreamContext->fragmentDescriptor.TimeScale, range))
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
40c46ba706b960f114341fbfddf61af2
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"range" is passed-by-value as parameter to "std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""range"")" instead of "range".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -8754,27 +8370,22 @@ void StreamAbstractionAAMP_MPD::FetchAndInjectInitialization(int trackIdx, bool
-												0.0, // duration - zero for init fragment
-												range.c_str(),
-												true ))
+										if(!FetchFragment(pMediaStreamContext, fragmentUrl, 0.0, true, getCurlInstanceByMediaType(pMediaStreamContext->mediaType), false, pMediaStreamContext->discontinuity, 0, pMediaStreamContext->fragmentDescriptor.TimeScale, range))
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
40c46ba706b960f114341fbfddf61af2
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"range" is passed-by-value as parameter to "std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""range"")" instead of "range".

</details>

rdkcmf-jenkins commented on dash/mpd/MPDModel.cpp:
```diff
@@ -179,6 +179,97 @@ string findBaseUrl(DomElement &element, const string &current, bool isFile) {
+                    for (const auto &base : current)
+                    {
+                        auto out = Url(base).resolve(newbase).format(Url::StripTrailingSlash).append(slash);
+                        baseUrls.push_back(out);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
60566bba01ee13fee5d40bb21908e10e
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"out" is copied and then passed-by-reference as parameter to STL insertion function "std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::push_back(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::value_type const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""out"")" instead of "out".

</details>

rdkcmf-jenkins commented on AampTrackWorkerManager.cpp:
```diff
@@ -0,0 +1,321 @@
+			std::shared_ptr<AampTrackWorker> worker = it->second;
+			if (worker)
+			{
+				return worker->SubmitJob(job, highPriority);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
ca351a5b7c54daa063439ce6b15b8560
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"job" is passed-by-value as parameter to "std::shared_ptr<aamp::AampTrackWorkerJob>::shared_ptr(std::shared_ptr<aamp::AampTrackWorkerJob> const &) /*explicit =default*/", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""job"")" instead of "job".

</details>

rdkcmf-jenkins commented on AampDownloadInfo.hpp:
```diff
@@ -0,0 +1,174 @@
+		  curlInstance(curlInstance),
+		  absolutePosition(absolutePosition),
+		  fragmentDurationSec(fragmentDurationSec),
+		  range(range),
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
9648ef2aa446d2b985dc9b80b0020151
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"range" is passed-by-value as parameter to "std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""range"")" instead of "range".

</details>

rdkcmf-jenkins commented on fragmentcollector_mpd.cpp:
```diff
@@ -10125,21 +9632,47 @@ void StreamAbstractionAAMP_MPD::FetcherLoop()
+							aamp->GetAampTrackWorkerManager()->WaitForCompletionWithTimeout(MAX_WAIT_TIMEOUT_MS, [this]() {
+								if (mIsLiveManifest)
+								{
+									UpdateMPD(); // Helps slide the progress window
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
166e3c9a19d5d07501a537c3551da637
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Unchecked return value
Calling "UpdateMPD" without checking return value (as is done elsewhere 4 out of 4 times).

_Medium Impact, CWE-252_
_CHECKED_RETURN_



</details>

rdkcmf-jenkins commented on AampTrackWorkerManager.cpp:
```diff
@@ -0,0 +1,321 @@
+				worker->ClearJobs();
+				// Submit a dummy job to ensure the worker is active and can process jobs
+				auto job = std::make_shared<AampTrackWorkerJob>();
+				auto future = worker->SubmitJob(job);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
820f638c07ef8a1da33dad5d956e05c5
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Variable copied when it could be moved
"job" is passed-by-value as parameter to "std::shared_ptr<aamp::AampTrackWorkerJob>::shared_ptr(std::shared_ptr<aamp::AampTrackWorkerJob> const &) /*explicit =default*/", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""job"")" instead of "job".

</details>

rdkcmf-jenkins commented on dash/mpd/MPDModel.cpp:
```diff
@@ -179,6 +179,97 @@ string findBaseUrl(DomElement &element, const string &current, bool isFile) {
+                {
+                    for (const auto &base : current)
+                    {
+                        auto out = Url(base).resolve(newbase).format(Url::StripTrailingSlash).append(slash);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
9be68a44798a02dd991e188ee01ed752
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Use of auto that causes a copy
Using the "auto" keyword without an "&" causes the copy of an object of type "std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >".

_Low Impact, CWE-none_
_AUTO_CAUSES_COPY_

## How to fix
 Use "std::move".

</details>

srikanthreddybijjam-comcast committed:
VPLAY-9274:[DASH] optimization: independent track downloads

Reason for change: Added Coverity Fixes.
Risks: Low
Test Procedure: Test with all sort of DASH contents
Priority: P1

Signed-off-by: srikanthreddybijjam-comcast <srikanthreddybijjam.2000@gmail.com>

rdkcmf-jenkins commented on dash/mpd/MPDModel.cpp:
```diff
@@ -179,6 +179,97 @@ string findBaseUrl(DomElement &element, const string &current, bool isFile) {
+                    for (const auto &base : current)
+                    {
+                        const auto& out = Url(base).resolve(newbase).format(Url::StripTrailingSlash).append(slash);
+                        baseUrls.push_back(std::move(out));
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
c20f7fa25952faeba3a0f8601ca02c2f
NOT_PRESENT
-->

Coverity issue no longer present as of: undefined
<details>
<summary>Show issue</summary>


# Coverity Issue - Pointer to local outside scope
Using "out", which points to an out-of-scope temporary variable of type "std::string".

_High Impact, CWE-562_
_RETURN_LOCAL_



</details>

pstroffolino commented on AampUtils.cpp:
```diff
@@ -1294,6 +1294,181 @@ bool parseAndValidateSCTE35(const std::string &scte35Data)
+ * @param toNumber number to replace token
+ * @retval position
+ */
+int replace(std::string &str, const std::string &from, uint64_t toNumber)
```
this is DASH-specific utility, might belong better in AampMPDUtils

nu641001 replied:
Done

pstroffolino commented on admanager_mpd.cpp:
```diff
@@ -1718,3 +1728,151 @@ bool PrivateCDAIObjectMPD::GetNextAdInBreakToPlace()
+ * @param[out] errorCode - AAMPCDAIError Error code if any.
+ * @return true if all init headers are fetched and cached successfully, false otherwise
+ */
+bool PrivateCDAIObjectMPD::FetchAndCacheInitHeaders(std::string& manifestStr, std::string& manifestUrl, AAMPCDAIError &errorCode)
```
Big function - is idea of collecting & caching all ad init headers up front new?  Why needed?

What happens if only some are available?  Will we refuse to play the ad?

nu641001 replied:
We attempt to download atleast one init fragment from each track to check whether ad is playable or not up front. Otherwise marked the ad as invalid

pstroffolino commented on AampTrackWorkerManager.cpp:
```diff
@@ -0,0 +1,329 @@
+ * limitations under the License.
+ */
+
+#include "AampTrackWorkerManager.hpp"
```
this seems a pretty generic class - the only thing track-specific about it is that MediaType is used as a key for each WorkerManager, yes?

nu641001 replied:
yes

pstroffolino commented on AampConfig.cpp:
```diff
@@ -447,6 +447,7 @@ static const ConfigLookupEntryInt mConfigLookupTableInt[AAMPCONFIG_INT_COUNT+CON
 	{0,"drmStallTimeout",eAAMPConfig_DrmStallTimeout,true,eCONFIG_RANGE_TIMEOUT},
 	{0,"drmStartTimeout",eAAMPConfig_DrmStartTimeout,true,eCONFIG_RANGE_TIMEOUT},
 	{0,"timeBasedBufferSeconds",eAAMPConfig_TimeBasedBufferSeconds,true,eCONFIG_RANGE_PLAYBACK_OFFSET},
+	{DEFAULT_MAX_DOWNLOAD_BUFFER,"maxDownloadBuffer",eAAMPConfig_MaxDownloadBuffer,true,eCONFIG_RANGE_PLAYBACK_OFFSET},
```
how does this work?  should I expect to see different download behavior/throttling in autotriage?

I'm seeing runaway downloads even with maxDownloadBuffer=4
Is this just a configuration for app-managed buffering in addition to gstreamers internal byte-based buffering?

nu641001 replied:
This config is right now used to control the queue based on a time based logic. This will move to a time based buffering logic with integration to with gstreamer position

nu641001 committed:
VPLAY-9274: [DASH] Dash specific utility APIs to AampMPDUtils

Reason for change: Moved Url replace utilities to AampMPDUtils from AampUtils
Risks: Low
Test Procedure: Test with DASH
Priority: P1

pstroffolino committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-9274

pstroffolino committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-9274

pstroffolino committed:
Update FakeAampTimeBasedBufferManager.cpp

Reason for Change: whitespace change to trigger new l1 test run
